runtime/callback.c

```diff
@@ -97,33 +137,79 @@ CAMLexport value caml_callback3_exn(value closure,
   return caml_callbackN_exn(closure, 3, arg);
 }
 
-#else
+#else /* Nativecode callbacks */
 
-/* Native-code callbacks. */
+static void init_callback_code(void)
+{
+}
 
-typedef value (callback_stub)(caml_domain_state* state, value closure,
+typedef value (callback_stub)(caml_domain_state* state,
+                              value closure,
                               value* args);
 
 callback_stub caml_callback_asm, caml_callback2_asm, caml_callback3_asm;
 
 CAMLexport value caml_callback_exn(value closure, value arg)
 {
-  return caml_callback_asm(Caml_state, closure, &arg);
+  caml_domain_state* domain_state = Caml_state;
+  caml_maybe_expand_stack();
+
+  if (Stack_parent(domain_state->current_stack)) {
+    CAMLparam2 (closure, arg);
+    CAMLlocal1 (cont);
+    value res;
+
+    cont = save_and_clear_stack_parent(domain_state);
+    res = caml_callback_asm(domain_state, closure, &arg);
+    restore_stack_parent(domain_state, cont);
+
+    CAMLreturn (res);
+  }
+  return caml_callback_asm(domain_state, closure, &arg);
 }
 
 CAMLexport value caml_callback2_exn(value closure, value arg1, value arg2)
 {
   value args[] = {arg1, arg2};
-  return caml_callback2_asm(Caml_state, closure, args);
+  caml_domain_state* domain_state = Caml_state;
+  caml_maybe_expand_stack();
+
+  if (Stack_parent(domain_state->current_stack)) {
+    CAMLparam3 (closure, arg1, arg2);
+    CAMLlocal1 (cont);
+    value res;
+
+    cont = save_and_clear_stack_parent(domain_state);
+    res = caml_callback2_asm(domain_state, closure, args);
+    restore_stack_parent(domain_state, cont);
+
+    CAMLreturn (res);
+  }
+  return caml_callback2_asm(domain_state, closure, args);
 }
 
 CAMLexport value caml_callback3_exn(value closure,
                                     value arg1, value arg2, value arg3)
 {
   value args[] = {arg1, arg2, arg3};
-  return caml_callback3_asm(Caml_state, closure, args);
+  caml_domain_state* domain_state = Caml_state;
+  caml_maybe_expand_stack();
+
+  if (Stack_parent(domain_state->current_stack))  {
+    CAMLparam4 (closure, arg1, arg2, arg3);
+    CAMLlocal1 (cont);
+    value res;
+
+    cont = save_and_clear_stack_parent(domain_state);
+    res = caml_callback3_asm(domain_state, closure, args);
+    restore_stack_parent(domain_state, cont);
+
+    CAMLreturn (res);
+  }
+  return caml_callback3_asm(domain_state, closure, args);
 }
 
+/* Native-code callbacks.  caml_callback[123]_exn are implemented in asm. */
```

Comment:
Is this what was intended?

```suggestion
/* Native-code callbacks.  caml_callback[123]_asm are implemented in asm. */
```

Reply:
Fixed in 54365f2